src/bin/pg_autoctl/fsm_transition.c

```diff
@@ -753,6 +753,103 @@ fsm_stop_postgres(Keeper *keeper)
 }
 
 
+/*
+ * fsm_stop_postgres_for_maintenance is used when pg_autoctl enable maintenance
+ * has been used on the primary server, we do a couple CHECKPOINT before
+ * stopping Postgres to ensure a smooth transition.
+ */
+bool
+fsm_stop_postgres_for_maintenance(Keeper *keeper)
+{
+	LocalPostgresServer *postgres = &(keeper->postgres);
+	PostgresSetup *pgSetup = &(postgres->postgresSetup);
+	PGSQL *pgsql = &(postgres->sqlClient);
+
+	if (pg_setup_is_running(pgSetup))
+	{
+		/*
+		 * PostgreSQL shutdown sequence includes a CHECKPOINT, that is issued
+		 * by the checkpointer process one every query backend has stopped
+		 * already. During this final CHECKPOINT no work can be done, so it's
+		 * best to reduce the amount of work needed there. To reduce the
+		 * checkpointer shutdown activity, we perform a manual shutdown while
+		 * still having concurrent activity.
+		 *
+		 * The first checkpoint writes all the in-memory buffers, the second
+		 * checkpoint writes everything that was added during the first one.
+		 */
+		log_info("Preparing shutdown for maintenance: CHECKPOINT;");
+
+		/* we may fail to CHECKPOINT and then will retry at next FSM loop */
+		for (int i = postgres->shutdownCheckpointCount; i < 2; i++)
+		{
+			if (!pgsql_checkpoint(pgsql))
+			{
+				log_error("Failed to checkpoint before stopping Postgres");
+				return false;
+			}
+
+			++(postgres->shutdownCheckpointCount);
+		}
+
+		/* cache invalidation */
+		postgres->shutdownCheckpointCount = 0;
+	}
+
+	return ensure_postgres_service_is_stopped(postgres);
+}
+
+
+/*
+ * fsm_stop_postgres_and_setup_standby is used when the primary is put to
+ * maintenance. Not only do we stop Postgres, we also prepare a partial setup
+ * as a secondary.
+ */
+bool
+fsm_stop_postgres_and_setup_standby(Keeper *keeper)
```

Comment:
The function name and/or comment don't indicate that this does the checkpoints before stopping or should be used for maintenance only.

Reply:
Added some comments about that.